rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,135 @@
+use crate::logger::{self, Level};
+use std::ffi::CString;
+use std::path::Path;
+use std::process;
+
+fn remove_lock(lock_filepath: &str) {
+    let path = CString::new(lock_filepath).expect("CString::new failed");
+    unsafe { libc::unlink(path.as_ptr()) };
+    log!(Level::Debug, "FsLock: removed lockfile {}", lock_filepath);
+}
+
+pub struct FsLock {
+    lock_filepath: String,
+    fd: i32,
+    locked: bool,
+}
+
+impl Default for FsLock {
+    fn default() -> Self {
+        FsLock {
+            lock_filepath: String::new(),
+            fd: -1,
+            locked: false,
+        }
+    }
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.locked {
+            remove_lock(&self.lock_filepath);
+            let _ = unsafe { libc::close(self.fd) };
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock<P: AsRef<Path>>(
+        &mut self,
+        new_lock_filepath: P,
+        pid: &mut libc::pid_t,
+    ) -> bool {
+        // switched to string as it is easier to work with c ffi later on
+        // Safety: this might fail if the path is not valid utf-8
+        let new_lock_filepath = new_lock_filepath.as_ref().to_str().unwrap();
+        if self.locked && self.lock_filepath == new_lock_filepath {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        // TODO: move to safe rust later
+        // let openoptions = OpenOptions::new().read(true).write(true).create(true);
+        // let mut file = match openoptions.open(path) {
+        //     Ok(file) => file,
+        //     Err(_) => return false,
+        // };
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_filepath
+        );
+
+        // first we open (and possibly create) the lock file
+        self.fd = unsafe {
+            libc::open(
+                new_lock_filepath.as_ptr() as *const i8,
```

Comment:
`as_ptr()` here crates a pointer to the first byte of Rust `String`, but `open()` expects a C string, i.e. one that is terminated by NUL. Since Rust's `String`s are not terminated that way, the `open()` call seemingly opens files that it shouldn't be able to.

Please construct a `CString` or `CStr`, and pass a pointer to that into `open()`.

Reply:
Yes, I just realized this while debugging.